components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/RemoteTabsStorage.kt

```diff
@@ -0,0 +1,149 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.storage.sync
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.withContext
+import mozilla.appservices.remotetabs.RemoteTab
+import mozilla.appservices.remotetabs.RemoteTabProviderException
+import mozilla.components.concept.storage.Storage
+import mozilla.components.concept.sync.SyncableStore
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.appservices.remotetabs.RemoteTabsProvider
+import mozilla.appservices.remotetabs.SyncAuthInfo as RustSyncAuthInfo
+import mozilla.components.concept.sync.SyncAuthInfo
+import mozilla.components.concept.sync.SyncStatus
+
+/**
+ * An interface which defines read/write methods for remote tabs data.
+ */
+open class RemoteTabsStorage : Storage, SyncableStore {
+    internal val api by lazy { RemoteTabsProvider() }
+    private val scope by lazy { CoroutineScope(Dispatchers.IO) }
+    internal val logger = Logger("PlacesHistoryStorage")
+
+    /**
+     * Store the locally opened tabs.
+     * @param tabs The list of opened tabs, for all opened non-private windows, on this device.
+     */
+    fun store(tabs: List<Tab>) {
+        api.setLocalTabs(tabs.map {
+            val activeTab = it.active()
+            val urlHistory = listOf(activeTab.url) + it.previous().reversed().map { it.url }
+            RemoteTab(activeTab.title, urlHistory, activeTab.iconUrl, it.lastUsed)
+        })
+    }
+
+    /**
+     * Get all remote devices tabs.
+     * @return A mapping of opened tabs per device.
+     */
+    fun getAll(): Map<Device, List<Tab>>? {
+        return api.getAll()?.map { Device.SyncDevice(it.clientId) to it.tabs.map {
```

Comment:
I think I'd rather keep the nulability of getAll() as we might not be synced yet

Reply:
(but I'll apply your suggestion of flow)

Reply:
Do consumers need to know that it's null because it's not yet synced? If so, let's make that explicit rather than use null as a way to determine it's not synced.

Reducing unnecessary nullability for consumers is better so that the number of null checks reduces as you go higher up the call stack. Returning an empty list is non-null way of saying there are no results.